File: cxx/constellation/controller/ControllerConfiguration.cpp
/**
 * @file
 * @brief Implementation of configuration parser
 *
 * @copyright Copyright (c) 2024 DESY and the Constellation authors.
 * This software is distributed under the terms of the EUPL-1.2 License, copied verbatim in the file "LICENSE.md".
 * SPDX-License-Identifier: EUPL-1.2
 */

#include "ControllerConfiguration.hpp"

#include <algorithm>
#include <cctype>
#include <filesystem>
#include <fstream>
#include <map>
#include <sstream>
#include <string>
#include <string_view>
#include <unordered_set>
#include <utility>
#include <variant>

#include <toml++/toml.hpp>
#include <yaml-cpp/yaml.h>

#include "constellation/controller/exceptions.hpp"
#include "constellation/controller/toml_helpers.hpp"
#include "constellation/controller/yaml_helpers.hpp"
#include "constellation/core/config/Configuration.hpp"
#include "constellation/core/config/value_types.hpp"
#include "constellation/core/log/log.hpp"
#include "constellation/core/protocol/CSCP_definitions.hpp"
#include "constellation/core/utils/string.hpp"

using namespace constellation::controller;
using namespace constellation::config;
using namespace constellation::log;
using namespace constellation::protocol;
using namespace constellation::utils;

ControllerConfiguration::ControllerConfiguration(const std::filesystem::path& path) {
    // Check if file exists
    if(!std::filesystem::is_regular_file(path)) {
        throw ConfigFileNotFoundError(path);
    }

    // Convert to absolute path
    const auto file_path_abs = std::filesystem::canonical(path);
    LOG(config_parser_logger_, DEBUG) << "Parsing configuration file " << quote(file_path_abs.string());

    const std::ifstream file {file_path_abs};
    std::ostringstream buffer {};
    buffer << file.rdbuf();

    const auto type = detect_config_type(path);
    switch(type) {
    case FileType::YAML: {
        parse_yaml(buffer.view());
        break;
    }
    case FileType::UNKNOWN:
    case FileType::TOML: {
        parse_toml(buffer.view());
        break;
    }
    default: std::unreachable();
    }

    // Validate the configuration
    validate();
}

ControllerConfiguration::ControllerConfiguration(std::string_view config, ControllerConfiguration::FileType type) {

    switch(type) {
    case FileType::YAML: {
        parse_yaml(config);
        break;
    }
    case FileType::UNKNOWN:
    case FileType::TOML: {
        parse_toml(config);
        break;
    }
    default: std::unreachable();
    }

    // Validate the configuration
    validate();
}

ControllerConfiguration::FileType ControllerConfiguration::detect_config_type(const std::filesystem::path& file) {
    const auto ext = transform(file.extension().string(), ::tolower);

    if(ext == ".yaml" || ext == ".yml") {
        return FileType::YAML;
    }
    if(ext == ".toml") {
        return FileType::TOML;
    }

    return FileType::UNKNOWN;
}

std::string ControllerConfiguration::getAsYAML() const {
    YAML::Emitter out;

    // Use flow-style for arrays
    out.SetSeqFormat(YAML::Flow);

    // Global dictionary
    out << YAML::BeginMap;

    // Add the global configuration keys
    if(!global_config_.empty()) {
        YAML::Node global_node {};
        for(const auto& [key, value] : global_config_) {
            global_node[key] = get_as_yaml_node(value);
        }
        out << YAML::Key << "_default";
        out << YAML::Value << global_node;
    }

    // Cache type nodes for later modification
    std::map<std::string, YAML::Node> type_nodes {};

    // Add type config
    for(const auto& [type, config] : type_configs_) {
        auto [type_node_it, inserted] = type_nodes.emplace(type, YAML::Node());
        auto& type_node = type_node_it->second;

        // Add type config keys
        YAML::Node node {};
        for(const auto& [key, value] : config) {
            node[key] = get_as_yaml_node(value);
        }
        type_node["_default"] = node;
    }

    // Append satellite configs to the type nodes:
    for(const auto& [canonical_name, config] : satellite_configs_) {
        const auto pos = canonical_name.find_first_of('.', 0);
        const auto type = canonical_name.substr(0, pos);
        const auto name = canonical_name.substr(pos + 1);

        auto [type_node_it, inserted] = type_nodes.emplace(type, YAML::Node());
        auto& type_node = type_node_it->second;

        // Add satellite config keys
        YAML::Node node {};
        for(const auto& [key, value] : config) {
            node[key] = get_as_yaml_node(value);
        }
        type_node[name] = node;
    }

    // Write final type nodes
    for(const auto& [key, node] : type_nodes) {
        out << YAML::Key << key;
        out << YAML::Value << node;
    }

    LOG_IF(config_parser_logger_, WARNING, !out.good()) << "Emitter error: " << out.GetLastError();

    // End global dictionary
    out << YAML::EndMap;
    return out.c_str();
}

std::string ControllerConfiguration::getAsTOML() const {
    // The global TOML table
    toml::table tbl {};
    tbl.emplace("_default", get_as_toml_table(global_config_));

    // Add type config
    for(const auto& [type, config] : type_configs_) {
        const auto [type_table_it, inserted] = tbl.emplace(type, toml::table());
        type_table_it->second.as_table()->emplace("_default", get_as_toml_table(config));
    }

    // Add config from individual satellites
    for(const auto& [canonical_name, config] : satellite_configs_) {
        const auto pos = canonical_name.find_first_of('.', 0);
        const auto type = canonical_name.substr(0, pos);
        const auto name = canonical_name.substr(pos + 1);
        const auto [type_table_it, inserted] = tbl.emplace(type, toml::table());
        type_table_it->second.as_table()->emplace(name, get_as_toml_table(config));
    }

    std::stringstream oss;
    oss << tbl << '\n';

    return oss.str();
}

void ControllerConfiguration::parse_yaml(std::string_view yaml) {
    YAML::Node root_node {};
    try {
        root_node = YAML::Load(std::string(yaml));
    } catch(const YAML::ParserException& e) {
        throw ConfigParseError(e.what());
    }

    // Root node needs to be a map or empty:
    if(!root_node.IsMap() && !root_node.IsNull()) {
        throw ConfigParseError("expected map as root node");
    }

    // Bool to check if global default config is defined multiple times
    bool has_global_default_config = false;

    // Loop over all nodes
    for(const auto& type_node_it : root_node) {
        const auto type_key_lc = parse_yaml_key(type_node_it);
        const auto& type_node = type_node_it.second;
        if(type_node.IsNull()) {
            // Skip if empty
            continue;
        }
        if(type_node.IsMap()) {
            if(type_key_lc == "_default") {
                // Global default config
                LOG(config_parser_logger_, DEBUG) << "Found default config at global level";
                if(has_global_default_config) {
                    throw ConfigKeyError(type_key_lc, "key defined twice");
                }
                global_config_ = parse_yaml_map(type_key_lc, type_node);
                has_global_default_config = true;
            } else {
                // Type level
                if(!CSCP::is_valid_satellite_type(type_key_lc)) {
                    throw ConfigKeyError(type_key_lc, "not a valid satellite type");
                }
                LOG(config_parser_logger_, DEBUG) << "Found type level for " << quote(type_key_lc);
                for(const auto& name_node_it : type_node) {
                    const auto name_key_lc = parse_yaml_key(name_node_it);
                    const auto canonical_name_key_lc = (type_key_lc + '.').append(name_key_lc);
                    const auto& name_node = name_node_it.second;
                    if(name_node.IsNull() && name_key_lc != "_default") {
                        // If node is empty, emplace empty satellite config and continue to next node
                        const auto [it, inserted] = satellite_configs_.try_emplace(canonical_name_key_lc);
                        if(!inserted) {
                            throw ConfigKeyError(canonical_name_key_lc, "key defined twice");
                        }
                        continue;
                    }
                    if(name_node.IsMap()) {
                        if(name_key_lc == "_default") {
                            // Type default config
                            LOG(config_parser_logger_, DEBUG)
                                << "Found default config at type level for " << quote(type_key_lc);
                            const auto [it, inserted] =
                                type_configs_.try_emplace(type_key_lc, parse_yaml_map(canonical_name_key_lc, name_node));
                            if(!inserted) {
                                throw ConfigKeyError(canonical_name_key_lc, "key defined twice");
                            }
                        } else {
                            // Satellite level
                            if(!CSCP::is_valid_satellite_name(name_key_lc)) {
                                throw ConfigKeyError(canonical_name_key_lc, "not a valid satellite name");
                            }
                            LOG(config_parser_logger_, DEBUG)
                                << "Found config at satellite level for " << quote(canonical_name_key_lc);
                            const auto [it, inserted] = satellite_configs_.try_emplace(
                                canonical_name_key_lc, parse_yaml_map(canonical_name_key_lc, name_node));
                            if(!inserted) {
                                throw ConfigKeyError(canonical_name_key_lc, "key defined twice");
                            }
                        }
                    } else {
                        throw ConfigValueError(canonical_name_key_lc, "expected a dictionary at satellite level");
                    }
                }
            }
        } else {
            throw ConfigValueError(type_key_lc, "expected a dictionary at type level");
        }
    }
}

void ControllerConfiguration::parse_toml(std::string_view toml) {
    toml::table tbl {};
    try {
        tbl = toml::parse(toml);
    } catch(const toml::parse_error& err) {
        std::ostringstream oss {};
        oss << err;
        throw ConfigParseError(oss.view());
    }

    // Bool to check if global default config is defined multiple times
    bool has_global_default_config = false;

    // Loop over all nodes
    tbl.for_each([this, &has_global_default_config](const toml::key& type_key, auto&& type_value) {
        const auto type_key_lc = transform(type_key.str(), ::tolower);
        if constexpr(toml::is_table<decltype(type_value)>) {
            if(type_key_lc == "_default") {
                // Global default config
                LOG(config_parser_logger_, DEBUG) << "Found default config at global level";
                if(has_global_default_config) {
                    throw ConfigKeyError(type_key_lc, "key defined twice");
                }
                global_config_ = parse_toml_table(type_key_lc, std::forward<decltype(type_value)>(type_value));
                has_global_default_config = true;
            } else {
                // Type level
                if(!CSCP::is_valid_satellite_type(type_key_lc)) {
                    throw ConfigKeyError(type_key_lc, "not a valid satellite type");
                }
                LOG(config_parser_logger_, DEBUG) << "Found type level for " << quote(type_key_lc);
                type_value.for_each([this, &type_key_lc](const toml::key& name_key, auto&& name_value) {
                    const auto name_key_lc = transform(name_key.str(), ::tolower);
                    const auto canonical_name_key_lc = (type_key_lc + '.').append(name_key_lc);
                    if constexpr(toml::is_table<decltype(name_value)>) {
                        if(name_key_lc == "_default") {
                            // Type default config
                            LOG(config_parser_logger_, DEBUG)
                                << "Found default config at type level for " << quote(type_key_lc);
                            const auto [it, inserted] = type_configs_.try_emplace(
                                type_key_lc,
                                parse_toml_table(canonical_name_key_lc, std::forward<decltype(name_value)>(name_value)));
                            if(!inserted) {
                                throw ConfigKeyError(canonical_name_key_lc, "key defined twice");
                            }
                        } else {
                            // Satellite level
                            if(!CSCP::is_valid_satellite_name(name_key_lc)) {
                                throw ConfigKeyError(canonical_name_key_lc, "not a valid satellite name");
                            }
                            LOG(config_parser_logger_, DEBUG)
                                << "Found config at satellite level for " << quote(canonical_name_key_lc);
                            const auto [it, inserted] = satellite_configs_.try_emplace(
                                canonical_name_key_lc,
                                parse_toml_table(canonical_name_key_lc, std::forward<decltype(name_value)>(name_value)));
                            if(!inserted) {
                                throw ConfigKeyError(canonical_name_key_lc, "key defined twice");
                            }
                        }
                    } else {
                        throw ConfigValueError(canonical_name_key_lc, "expected a dictionary at satellite level");
                    }
                });
            }
        } else {
            throw ConfigValueError(type_key_lc, "expected a dictionary at type level");
        }
    });
}

void ControllerConfiguration::fill_dependency_graph(std::string_view canonical_name) {
    // Parse all transition condition parameters
    for(const auto& state : {CSCP::State::initializing,
                             CSCP::State::launching,
                             CSCP::State::landing,
                             CSCP::State::starting,
                             CSCP::State::stopping}) {
        const auto key = transform("_require_" + to_string(state) + "_after", ::tolower);

        // Get final assembled config and look for the key:
        const auto config = getSatelliteConfiguration(canonical_name);
        if(config.contains(key)) {
            auto register_dependency = [this, canonical_name, state](auto& dependent) {
                // Register dependency in graph, current satellite depends on config value satellite
                transition_graph_[state][transform(dependent, ::tolower)].emplace(transform(canonical_name, ::tolower));
            };

            const auto dependents = Configuration(config).getArray<std::string>(key);
            LOG(config_parser_logger_, DEBUG)
                << "Registering dependency for transitional state " << quote(to_string(state)) << " of " << canonical_name
                << " with dependents " << range_to_string(dependents);
            std::ranges::for_each(dependents, register_dependency);
        }
    }
}

void ControllerConfiguration::validate() {
    // Build the dependency graph from all satellite configurations
    for(const auto& [name, cfg] : satellite_configs_) {
        fill_dependency_graph(name);
    }

    // Check each transition for possible cycles
    for(const auto& transition_pair : transition_graph_) {
        const auto transition = transition_pair.first;
        LOG(config_parser_logger_, DEBUG) << "Checking for deadlock in transition: " << transition;

        if(check_transition_deadlock(transition)) {
            LOG(config_parser_logger_, DEBUG) << "Deadlock detected in transition: " << transition;
            throw ConfigValidationError("Cyclic dependency for transition " + quote(to_string(transition)));
        }
    }
    // No deadlock in any transition
}

bool ControllerConfiguration::check_transition_deadlock(CSCP::State transition) const {
    // If no dependencies for this transition doesn't exist, there is no cycle:
    if(!transition_graph_.contains(transition)) {
        return false;
    }

    std::unordered_set<std::string> visited;
    std::unordered_set<std::string> recursion_stack;

    // Recursive depth first search:
    auto dfs = [&](const auto& self, const std::string& satellite) { // NOLINT(misc-no-recursion)
        // Cycle detected (deadlock)
        if(recursion_stack.contains(satellite)) {
            return true;
        }

        // Satellite already processed
        if(visited.contains(satellite)) {
            return false;
        }

        // No transition registered for this satellite:
        if(!transition_graph_.at(transition).contains(satellite)) {
            return false;
        }

        visited.insert(satellite);
        recursion_stack.insert(satellite);

        // Visit all dependent satellites
        for(const auto& dependent : transition_graph_.at(transition).at(satellite)) {
            if(self(self, dependent)) {
                return true;
            }
        }

        // Remove satellite from recursion stack
        recursion_stack.erase(satellite);
        return false;
    };

    // Traverse each satellite for the given transition
    const auto deadlock = std::ranges::any_of(transition_graph_.at(transition), [&](const auto& p) {
        const std::string& satellite = p.first;
        if(!visited.contains(satellite)) {
            if(dfs(dfs, satellite)) {
                // Deadlock detected in this transition
                return true;
            }
        }
        return false;
    });

    return deadlock;
}

// NOLINTNEXTLINE(misc-no-recursion)
void ControllerConfiguration::overwrite_config(const std::string& key_prefix,
                                               Dictionary& base_config,
                                               const Dictionary& config) const {
    for(const auto& [key, value] : config) {
        auto base_kv_it = base_config.find(key);
        if(base_kv_it != base_config.cend()) {
            // Overwrite existing parameter
            const auto prefixed_key = key_prefix + key;
            auto& base_value = base_kv_it->second;
            // Ensure that dictionary types match
            if((std::holds_alternative<Dictionary>(value) || std::holds_alternative<Dictionary>(base_value)) &&
               (value.index() != base_value.index())) {
                throw ConfigValidationError("value of key " + quote(prefixed_key) +
                                            " has mismatched types when merging defaults");
            }
            if(std::holds_alternative<Dictionary>(base_value)) {
                // If dictionary, overwrite recursively
                overwrite_config(prefixed_key + '.', base_value.get<Dictionary>(), value.get<Dictionary>());
            } else {
                // Otherwise overwrite directly (ignore type mismatch)
                base_value = value;
                LOG(config_parser_logger_, TRACE) << "Overwritten value for key " << quote(prefixed_key);
            }
        } else {
            // Add new parameter
            base_config.emplace(key, value);
        }
    }
}

bool ControllerConfiguration::hasTypeConfiguration(std::string_view type) const {
    return type_configs_.contains(transform(type, ::tolower));
}

void ControllerConfiguration::addTypeConfiguration(std::string_view type, Dictionary config) {
    // Check if already there
    const auto type_lc = transform(type, ::tolower);
    auto config_it = type_configs_.find(type_lc);
    if(config_it != type_configs_.cend()) {
        LOG(config_parser_logger_, WARNING) << "Overwriting existing satellite type configuration for " << quote(type);
        config_it->second = std::move(config);
    } else {
        type_configs_.emplace(type_lc, std::move(config));
    }
}

Dictionary ControllerConfiguration::getTypeConfiguration(std::string_view type) const {
    LOG(config_parser_logger_, TRACE) << "Fetching configuration for type " << quote(type);

    // Copy global config
    auto config = getGlobalConfiguration();

    // Add parameters from type level
    const auto type_lc = transform(type, ::tolower);
    const auto config_it = type_configs_.find(transform(type_lc, ::tolower));
    if(config_it != type_configs_.end()) {
        LOG(config_parser_logger_, TRACE) << "Found config at type level for " << quote(type);
        overwrite_config("", config, config_it->second);
    }

    return config;
}

bool ControllerConfiguration::hasSatelliteConfiguration(std::string_view canonical_name) const {
    return satellite_configs_.contains(transform(canonical_name, ::tolower));
}

void ControllerConfiguration::addSatelliteConfiguration(std::string_view canonical_name, Dictionary config) {
    // Check if already there
    const auto canonical_name_lc = transform(canonical_name, ::tolower);
    auto config_it = satellite_configs_.find(canonical_name_lc);
    if(config_it != satellite_configs_.end()) {
        LOG(config_parser_logger_, WARNING) << "Overwriting existing satellite configuration for " << quote(canonical_name);
        config_it->second = std::move(config);
    } else {
        satellite_configs_.emplace(canonical_name_lc, std::move(config));
    }
}

Dictionary ControllerConfiguration::getSatelliteConfiguration(std::string_view canonical_name) const {
    LOG(config_parser_logger_, TRACE) << "Fetching configuration for " << quote(canonical_name);

    // Find type from canonical name
    const auto canonical_name_lc = transform(canonical_name, ::tolower);
    const auto separator_pos = canonical_name_lc.find_first_of('.');
    const auto type_lc = canonical_name_lc.substr(0, separator_pos);

    // Copy from global global + type level
    auto config = getTypeConfiguration(type_lc);

    // Add parameters from satellite level
    const auto config_it = satellite_configs_.find(canonical_name_lc);
    if(config_it != satellite_configs_.end()) {
        LOG(config_parser_logger_, TRACE) << "Found config at satellite level for " << quote(canonical_name);
        overwrite_config("", config, config_it->second);
    }

    return config;
}

File: python/constellation/core/controller_configuration.py
"""
SPDX-FileCopyrightText: 2025 DESY and the Constellation authors
SPDX-License-Identifier: EUPL-1.2

Controller configuration class
"""

from __future__ import annotations

import copy
import enum
import pathlib
import tomllib
from typing import Any

import yaml

from constellation.core.configuration import Configuration


class ConfigParseError(RuntimeError):
    def __init__(self, reason: str):
        super().__init__(f"Could not parse configuration: {reason}")


class ConfigKeyError(ConfigParseError):
    def __init__(self, key: str, reason: str):
        super().__init__(f"error while parsing key `{key}`: {reason}")


class ConfigValueError(ConfigParseError):
    def __init__(self, key: str, reason: str):
        super().__init__(f"error while parsing value of key `{key}`: {reason}")


class ConfigValidationError(ConfigParseError):
    def __init__(self, reason: str):
        super().__init__(f"error validating configuration: {reason}")


def _convert_to_config(config: Configuration | dict[str, Any]) -> Configuration:
    if not isinstance(config, Configuration):
        try:
            return Configuration(config)
        except Exception as e:
            raise ConfigParseError(str(e)) from e
    return config


def _override_dict(base_dict: dict[str, Any], new_dict: dict[str, Any], prefix: str = "") -> None:
    for key, value in new_dict.items():
        if key in base_dict:
            # Overwrite existing parameter
            prefixed_key = prefix + key
            base_value = base_dict[key]
            value_is_dict = isinstance(value, dict)
            base_value_is_dict = isinstance(base_value, dict)
            if value_is_dict or base_value_is_dict:
                if value_is_dict != base_value_is_dict:
                    raise ConfigValidationError(f"value of key `{prefixed_key}` has mismatched types when merging defaults")
                # If dictionary, overwrite recursively
                _override_dict(base_value, value, prefixed_key + ".")
            else:
                # Otherwise overwrite directly (ignore type mismatch)
                base_dict[key] = value
        else:
            # Add new parameter
            base_dict[key] = value


class FileType(enum.Enum):
    UNKNOWN = enum.auto()
    TOML = enum.auto()
    YAML = enum.auto()


class ControllerConfiguration:
    def __init__(self) -> None:
        self._global_config = Configuration()
        self._type_configs = dict[str, Configuration]()
        self._satellite_configs = dict[str, Configuration]()

    def __str__(self) -> str:
        return str(self._as_single_dict())

    @staticmethod
    def from_string(config_string: str, file_type: FileType) -> ControllerConfiguration:
        """Create a controller configuration from a string in TOML or YAML"""
        config = ControllerConfiguration()
        if file_type == FileType.YAML:
            config._parse_yaml(config_string)
        else:
            # Assume UNKNOWN is TOML
            config._parse_toml(config_string)
        return config

    @staticmethod
    def from_path(path: pathlib.Path) -> ControllerConfiguration:
        """Create a controller configuration from a file"""
        file_type = FileType.UNKNOWN
        suffix = path.suffix
        if suffix == ".toml":
            file_type = FileType.TOML
        elif suffix in [".yaml", ".yml"]:
            file_type = FileType.YAML
        with open(path) as file:
            return ControllerConfiguration.from_string(file.read(), file_type)

    def set_global_configuration(self, config: Configuration | dict[str, Any]) -> None:
        """Set the global configuration section"""
        self._global_config = _convert_to_config(config)

    def get_global_configuration(self) -> Configuration:
        """Get the global configuration section"""
        return self._global_config

    def has_type_configuration(self, satellite_type: str) -> bool:
        """Check if an explicit configuration exists for a given satellite type"""
        return satellite_type.lower() in self._type_configs

    def add_type_configuration(self, satellite_type: str, config: Configuration | dict[str, Any]) -> None:
        """Add or override an explicit configuration for a satellite type"""
        self._type_configs[satellite_type.lower()] = _convert_to_config(config)

    def get_type_configuration(self, satellite_type: str) -> Configuration:
        """Get configuration for a given satellite type"""
        satellite_type_lc = satellite_type.lower()

        # Copy global config
        config = Configuration(copy.deepcopy(self._global_config._dictionary))

        # Add parameters from type level
        if satellite_type_lc in self._type_configs:
            _override_dict(config._dictionary, self._type_configs[satellite_type_lc]._dictionary)

        return config

    def has_satellite_configuration(self, canonical_name: str) -> bool:
        """Check if an explicit configuration exists for a given satellite"""
        return canonical_name.lower() in self._satellite_configs

    def add_satellite_configuration(self, canonical_name: str, config: Configuration | dict[str, Any]) -> None:
        """Add or override an explicit configuration for a satellite"""
        self._satellite_configs[canonical_name.lower()] = _convert_to_config(config)

    def get_satellite_configuration(self, canonical_name: str) -> Configuration:
        """Get configuration for a given satellite"""
        canonical_name_lc = canonical_name.lower()

        # Find type from canonical name
        satellite_type_lc = canonical_name_lc.split(".")[0]

        # Get configuration from global and type level
        config = self.get_type_configuration(satellite_type_lc)

        # Add parameters from satellite level
        if canonical_name_lc in self._satellite_configs:
            _override_dict(config._dictionary, self._satellite_configs[canonical_name_lc]._dictionary)

        return config

    def _parse_dict(self, config_dict: dict[Any, Any]) -> None:
        has_global_default_config = False
        for type_key, type_value in config_dict.items():
            if not isinstance(type_key, str):
                raise ConfigKeyError(str(type_key), "key not a string")
            type_key_lc = type_key.lower()

            if not isinstance(type_value, dict):
                raise ConfigValueError(type_key_lc, "expected a dictionary at type level")

            if type_key_lc == "_default":
                # Global default config
                if has_global_default_config:
                    raise ConfigKeyError(type_key_lc, "key defined twice")
                self.set_global_configuration(type_value)
                has_global_default_config = True
            else:
                # Type level
                for name_key, name_value in type_value.items():
                    if not isinstance(name_key, str):
                        raise ConfigKeyError(f"{type_key_lc}.{name_key}", "key not a string")
                    name_key_lc = name_key.lower()
                    canonical_name_lc = f"{type_key_lc}.{name_key_lc}"

                    if not isinstance(name_value, dict):
                        raise ConfigValueError(canonical_name_lc, "expected a dictionary at satellite level")

                    if name_key_lc == "_default":
                        # Type default config
                        if self.has_type_configuration(type_key_lc):
                            raise ConfigKeyError(canonical_name_lc, "key defined twice")
                        self.add_type_configuration(type_key_lc, name_value)
                    else:
                        # Satellite level
                        if self.has_satellite_configuration(canonical_name_lc):
                            raise ConfigKeyError(canonical_name_lc, "key defined twice")
                        self.add_satellite_configuration(canonical_name_lc, name_value)

    def _parse_toml(self, config_string: str) -> None:
        toml_dict: dict[str, Any]
        try:
            toml_dict = tomllib.loads(config_string)
        except Exception as e:
            raise ConfigParseError(str(e)) from e
        self._parse_dict(toml_dict)

    def _parse_yaml(self, config_string: str) -> None:
        # Workaround for YAML
        def adjust_parsed_yaml_dict(yaml_dict: dict[Any, Any]) -> None:
            for key, value in yaml_dict.items():
                if value is None:
                    # Interpret empty node as empty dictionary
                    yaml_dict[key] = {}
                elif isinstance(value, dict):
                    # Run recursively
                    adjust_parsed_yaml_dict(value)

        yaml_dict: Any
        try:
            yaml_dict = yaml.safe_load(config_string)
        except Exception as e:
            raise ConfigParseError(str(e)) from e
        if yaml_dict is None:
            yaml_dict = {}
        if not isinstance(yaml_dict, dict):
            raise ConfigParseError("expected map as root node")
        adjust_parsed_yaml_dict(yaml_dict)
        self._parse_dict(yaml_dict)

    def _as_single_dict(self) -> dict[str, Any]:
        single_dict = {"_default": self._global_config._dictionary}
        for key, value in self._type_configs.items():
            single_dict[key] = {"_default": value._dictionary}
        for key, value in self._satellite_configs.items():
            satellite_type, satellite_name = key.split(".")
            if satellite_type not in single_dict:
                single_dict[satellite_type] = {}
            single_dict[satellite_type][satellite_name] = value._dictionary
        return single_dict


def load_config(path: str | pathlib.Path) -> ControllerConfiguration:
    """Load a configuration file"""
    if isinstance(path, str):
        path = pathlib.Path(path)
    return ControllerConfiguration.from_path(path)
